nexus/saga-recovery/src/lib.rs

```diff
@@ -0,0 +1,659 @@
+    // of this crate.
+    //
+    // Tests the following structures used together:
+    //
+    // - RestState
+    // - Plan
+    // - Execution
+    // - Report
+    //
+    // These are hard to test in isolation since they're intended to be used
+    // together in a loop (and so don't export public interfaces for mucking
+    // with internal).
+    #[tokio::test]
+    async fn test_basic() {
+        let logctx = test_setup_log("saga_recovery_basic");
+        let log = &logctx.log;
+
+        // Start with a blank slate.
+        let mut sim = Simulator::new(log.clone());
+        let initial_rest_state = sim.rest_state.clone();
+        let mut report = status::Report::new();
+
+        //
+        // Now, go through a no-op recovery.
+        //
+        let (plan, execution, last_pass_result, nstarted) =
+            sim.sim_recovery_pass();
+        assert_eq!(last_pass_result.nfound, 0);
+        assert_eq!(last_pass_result.nskipped, 0);
+        assert_eq!(last_pass_result.nremoved, 0);
+        assert_eq!(sim.rest_state, initial_rest_state);
+        report.update_after_pass(&plan, execution, nstarted);
+        assert_eq!(report.ntotal_recovered, 0);
+        assert_eq!(report.ntotal_failures, 0);
+        assert_eq!(report.ntotal_started, 0);
+        assert_eq!(report.ntotal_finished, 0);
+
+        //
+        // Now, go through a somewhat general case of recovery.
+        //
+        // First, add a couple of sagas that just showed up in the database.
+        // This covers the case of sagas that were either from a previous Nexus
+        // lifetime or re-assigned from some other Nexus that has been expunged.
+        // We create two so we can exercise success and failure cases for
+        // recovery.
+        //
+        let saga_recover_ok = sim.sim_previously_running_saga();
+        let saga_recover_fail = sim.sim_previously_running_saga();
+        sim.sim_config_recovery_result(saga_recover_fail, true);
+
+        // Simulate Nexus starting a couple of sagas in the usual way.  This one
+        // will appear in the database as well as in our set of sagas started.
+        let saga_started_normally_1 = sim.sim_normal_saga_start();
+        let saga_started_normally_2 = sim.sim_normal_saga_start();
+
+        // Start a saga and then finish it immediately.  This is a tricky case
+        // because the recovery pass will see that it started, but not see in
+        // the database, and it won't be able to tell if it finished or just
+        // started.
+        let saga_started_and_finished = sim.sim_normal_saga_start();
+        sim.sim_normal_saga_done(saga_started_and_finished);
+
+        // Take a snapshot.  Subsequent changes will not affect the database
+        // state that's used for the next recovery pass.  We'll use this to
+        // simulate Nexus having started a saga immediately after the database
+        // listing that's used for a recovery pass.
```

Comment:
I added some println logging to the simulator and some overview stuff.  I didn't add super detailed descriptions of what it was doing because that seemed more annoying to keep up to date.